test/SILGen/capture-transitive.swift

```diff
@@ -0,0 +1,20 @@
+// RUN: %target-swift-emit-silgen -enable-sil-ownership %s | %FileCheck %s
+// SR-8398
+func fibonacci(_ n: Int) -> Int {
+  var cache: [Int: Int] = [:]
+
+  func recursive(_ m: Int) -> Int {
+    return cache[m] ?? {
+      // Make sure cache is only captured once in the closure
+      // CHECK: implicit closure #1 in recursive #1
+      // CHECK-LABEL: sil private [transparent] @{{.*}}9fibonacci{{.*}}9recursive{{.*}} : $@convention(thin) (Int, @guaranteed { var Dictionary<Int, Int> }) -> (Int, @error Error)
+      // CHECK: closure #1 in implicit closure #1 in recursive #1
+      // CHECK-LABEL: sil private @{{.*}}9fibonacci{{.*}}9recursive{{.*}} : $@convention(thin) (Int, @guaranteed { var Dictionary<Int, Int> }) -> Int
+      let output = m < 2 ? m : recursive(m - 1) + recursive(m - 2)
```

Comment:
Can you add tests for weak and unowned as well?

Reply:
If we make a similar closure to your first example, we can capture `self` strong two different ways in the closure:
```
func myGeneric<T>(_: T) -> T.Type {
    return T.self
}

class C {
    func returnsSelf() -> Self {
      func f() {
        print(myGeneric(self)) // we're calling myGeneric() with T := DynamicSelfType
      }
      {
        f()
        _ = myGeneric(self)
      }()
      return self
    }
}
```
`f()` captures `self` with `IsDirect` and the unnamed closure captures `self` with `IsDirect|IsNoEscape` and transitively also needs to deal with `f()`s capture. (Side note: I haven't investigated why `IsNoEscape` is set for one and not the other.) So merging flags is appropriate when dealing with multiple explicit captures of `self`.

But then if there are no explicit captures of `self`, then we get to your second example, where none of this code gets exercised at all, and the `selfCapture` optional remains empty. Finally, the dynamic self metadata gets added to the tail end of the lowered captures for use later. But at this level, no other function/closure sees that here, because the transitiveness is dealing only with _explicit_ captures. So there's no point at which that implicit metadata capture gets compared to anything else or has its flags (which are created `0`) checked.